# Aircontrol Input API

Aircontrol provides fine control over inputs via python API. 
Following properties can be controlled using `set_control` function.

## Python API
- InputControlType (str, optional): It can be either `Code` or `Other`. This is to control the internal mechanism and prevent repeted calling in already set variables. \
    If `InputControlType` is set to 'Code', camera cannot be controlled from Keyboard or Joystick. If `InputControlType` is set to 'Other', camera can be only controlled from Keyboard or Joystick.  Defaults to "Code". 
- Pitch (float, optional): The aircraft nose can rotate up and down about the y-axis, a motion known as pitch. Pitch control is typically accomplished using an elevator on the horizontal tail. Defaults to 0.0.
- Roll (float, optional): The wingtips can rotate up and down about the x-axis, a motion known as roll. Roll control is usually provided using ailerons located at each wingtip.. Defaults to 0.0.
- Yaw (float, optional): The nose can rotate left and right about the z-axis, a motion known as yaw.Yaw control is most often accomplished using a rudder located on the vertical tail.. Defaults to 0.0.
- Throttle (float, optional): Controls the engine power (Deprecated), Use sticky Throttle. Defaults to 0.0.
- StickyThrottle (float, optional): Tyoically Airplane have sticky throttles. Throttle values stay same unless moved. Defaults to 0.0.
- Brake (int, optional): Applies brake to wheels, to control ground movement. Defaults to 0.
- Flaps (int, optional): A flap is a high-lift device used to reduce the stalling speed of an aircraft wing at a given weight. Flaps helps in controlling descent. Defaults to 0.

## Limits
| Control  | type  | Limit |
|---|---|---|
| Pitch  | float | Between -1.0 and 1.0 |
| Yaw  | float | Between -1.0 and 1.0 |
| Roll  | float | Between -1.0 and 1.0 |
| Sticky Throttle  | float  |  Between 0.0 and 1.0 | |
| Brake  | int  | 0 : Dis-engaged, 1 : Engaged  |
| Flaps  | int | 0 : None, 1 : 15 Degree, 2 : 30 Degree  |

## Importing Requirements

In [1]:
from airctrl import environment
from pprint import pprint
import PIL.Image as Image
import base64
import numpy as np
from io import BytesIO
from matplotlib.pyplot import  imshow
import matplotlib.pyplot as plt

In [2]:
from airctrl import environment 
from airctrl import sample_generator
from airctrl.utils import unity
from airctrl import sample_generator
port=8999
sample = sample_generator.samples()
env =  environment.Trigger()
L = unity.Launch()

Now call method `.get_connected(port=<Default 8053>)` to get connected



In [3]:
process = L.launch_executable("/home/supatel/Games/AirControl_2021/Build/1.3.0/Linux/v1.3.0-AirControl.x86_64", server_port=port)
env.get_connected(port=port)

Loading environment from /home/supatel/Games/AirControl_2021/Build/1.3.0/Linux/v1.3.0-AirControl.x86_64 at port 8999 client ip 127.0.1.1 client port 8999

['/home/supatel/Games/AirControl_2021/Build/1.3.0/Linux/v1.3.0-AirControl.x86_64', '--serverPort', '8999', '--clientIP', '127.0.1.1', '--clientPort', '8999']
Sleeping for 5 seconds to allow environment load

[UnityMemory] Configuration Parameters - Can be set up in boot.config
    "memorysetup-bucket-allocator-granularity=16"
    "memorysetup-bucket-allocator-bucket-count=8"
    "memorysetup-bucket-allocator-block-size=4194304"
    "memorysetup-bucket-allocator-block-count=1"
    "memorysetup-main-allocator-block-size=16777216"
    "memorysetup-thread-allocator-block-size=16777216"
    "memorysetup-gfx-main-allocator-block-size=16777216"
    "memorysetup-gfx-thread-allocator-block-size=16777216"
    "memorysetup-cache-allocator-block-size=4194304"
    "memorysetup-typetree-allocator-block-size=2097152"
    "memorysetup-profiler-bucke

In [4]:
control_schema = env.step(Pitch=0.0, Roll=0.0, Yaw=0.0,Throttle=0.0, StickyThrottle=0.5, Brake=0, Flaps=0,IsOutput=True)
print(control_schema)

{'AGL': 0.0, 'MSL': 4.7329154, 'CurrentRPM': 0.0, 'MaxRPM': 4500.0, 'MaxPower': 1.10416677e-05, 'CurrentPower': 0.0, 'CurrentFuel': 0.0, 'CurrentSpeed': 0.0, 'BankAngle': 0.0017182061, 'IfCollision': False, 'CollisionObject': '', 'Latitude': -1.70281146e-06, 'Longitude': 5.317519e-06, 'PitchAngle': 0.328555942, 'ScreenCapture': '', 'LidarPointCloud': [500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 500.0, 30

## Output
When `IsOutput` is set to `True`, the Aircontrol reply with the following output.

```
{
   "AGL":0.0,
   "MSL":4.7329154,
   "CurrentRPM":0.0,
   "MaxRPM":4500.0,
   "MaxPower":1.10416677e-05,
   "CurrentPower":0.0,
   "CurrentFuel":0.0,
   "CurrentSpeed":0.0,
   "BankAngle":0.0017182061,
   "IfCollision":false,
   "CollisionObject":"",
   "Latitude":-1.70281146e-06,
   "Longitude":5.317519e-06,
   "PitchAngle":0.328555942,
   "ScreenCapture":"Here the screen capture will be returned as base64 o",
   "LidarPointCloud":[
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      305.198425,
      190.873642,
      138.888733,
      109.184929,
      89.97085,
      76.52718,
      66.596344,
      58.9627533,
      52.91375,
      48.003334,
      43.93948,
      40.5215073,
      37.60714,
      35.0940552,
      32.9049,
      30.9822521,
      29.2800446,
      27.7629089,
      26.4031944,
      25.1775265,
      24.0678177,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      500.0,
      ...,
      ...,
      ...,
   ],
   "Counter":0,
   "log":"",
   "Reward":0.0035562878254627784,
   "IsGrounded":true,
   "IsFlying":false,
   "IsTaxiing":false,
   "PosXAbs":0.86777246,
   "PosYAbs":1.44259226,
   "PosZAbs":-0.0676386356,
   "PosXRel":-0.00222843885,
   "PosYRel":-0.327407122,
   "PosZRel":-0.09763813,
   "RotXAbs":340.814636,
   "RotYAbs":359.966278,
   "RotZAbs":0.105423734,
   "RotXRel":340.814636,
   "RotYRel":359.966278,
   "RotZRel":0.105423853,
   "AngularXVelocity":-0.00521957874,
   "AngularYVelocity":-8.102572e-05,
   "AngularZVelocity":0.00138697785,
   "LinearXVelocity":-0.00184028456,
   "LinearYVelocity":0.007154375,
   "LinearZVelocity":-0.09623016,
   "AngularXAcceleration":5.92379365e-05,
   "AngularYAcceleration":-0.0001057535,
   "AngularZAcceleration":0.0,
   "LinearXAcceleration":-0.000308157178,
   "LinearYAcceleration":-0.030831784,
   "LinearZAcceleration":-0.18662475,
   "MsgType":"Output",
   "Version":"1.3.0"
}
```

![Bank Angle](../../docs/images/bankAngle.png)

Angle between vertical component of the Airplane w.r.t global vertical.[Source](https://en.wikipedia.org/wiki/Flight_dynamics_(fixed-wing_aircraft))

![Pitch Angle](../../docs/images/pitchAngle.png)

Angle between vertical component of the Airplane w.r.t global vertical.[Source](https://en.wikipedia.org/wiki/Flight_dynamics_(fixed-wing_aircraft))

|Measure|Description|example|
|:---:|:---:|:---:|
|AGL|Above Ground Level, or AGL, describes the literal height above the ground over which you’re flying.This also accounts for the building and mounntain heights.| 0.0|
|MSL|Mean Sea Level, or MSL, is your true altitude or elevation. This is measured as differrence of current height and 0.0 in XZ  plane.| 4.7329154|
|BankAngle| The roll angle is also known as bank angle on a fixed-wing aircraft, which usually "banks" to change the horizontal direction of flight. Value in radian limit -1 to +1.| 0.0|
|PitchAngle| Angle between the aircraft longitudinal axis and horizontal; Value in radian limit -1 to +1.| 0.328555942|
|CurrentFuel| [WIP] : Current fuel in gallons.| 29.4|
|CurrentPower| Current engine power.| 1.10416677e-05|
|Current RPM| Current Engine RPM.| 0.0|
|CurrentSpeed| Current Airplane Speed in Knots.|
|MaxRPM| : Max engine RPM.| 4500.0|
|ScreenCapture| Captured Screenshot, if proper trigger set from `set_camera` function|Base 64 Image|
|LidarPointCloud| Captured Lidar point cloud, if proper trigger set from `set_lidar` function| Vector : [500.0, 500.0, 500.0, ..., 25, 190.873642, 138.888733, 109.184929]|
|Counter| Number of request the client has sent to the server.| 13|
|log|: Log of events, if requested.| |
|Reward| Internal example reward system. user can define their own reward system externally.| 0.0035562878254627784|
|IsGrounded| Indicates if the Airplane is grounded.| true|
|IsFlying| Indicates if the Airplane is flying.| false|
|IsTaxiing| Indicates if the Airplane is taxing on runway.| false|
|PosXAbs| Absolute position on X axis.| 0.86777246|
|PosYAbs| Absolute position on Y axis. | 1.44259226|
|PosZAbs| Absolute position on Z axis. | -0.0676386356|
|PosXRel| Relative position w.r.t start position of the Airplane on X axis.| -0.00222843885|
|PosYRel| Absolute position w.r.t start position of the Airplane on Y axis.| -0.327407122|
|PosZRel| Absolute position w.r.t start position of the Airplane on Z axis.| -0.09763813|
|RotXAbs| Absolute rotation on X axis.| 340.814636|
|RotYAbs| Absolute rotation on Y axis. | 359.966278|
|RotZAbs| Absolute rotation on Z axis. | 0.105423734|
|RotXRel| Absolute rotation w.r.t start rotation of the Airplane on Y axis.| 340.814636|
|RotYRel| Absolute rotation w.r.t start rotation of the Airplane on Y axis.| 359.966278|
|RotZRel| Absolute position w.r.t start rotation of the Airplane on Y axis.| 0.105423853|
|AngularXVelocity| Angular velocity on X axis.| -0.00521957874|
|AngularYVelocity| Angular velocity on Y axis.| -8.102572e-05|
|AngularZVelocity| Angular velocity on Z axis.| 0.00138697785|
|LinearXVelocity| Linear velocity on X axis.| -0.00184028456|
|LinearYVelocity| Linear velocity on Y axis.| 0.007154375|
|LinearZVelocity| Linear velocity on Z axis.| -0.09623016|
|AngularXAcceleration| Angular acceleration on X axis.| 5.92379365e-05|
|AngularYAcceleration| Angular acceleration on Y axis.| -0.0001057535|
|AngularZAcceleration| Angular acceleration on Z axis.| 0.0|
|LinearXAcceleration| Linear acceleration on Z axis.| -0.000308157178|
|LinearYAcceleration| Linear acceleration on Z axis.| -0.030831784|
|LinearZAcceleration| Linear acceleration on Z axis. |-0.18662475|
|MsgType|  Message type will be output. Example |Output|
|Version| Version of the server. |1.3.0|